api/equinixmetal: do not exclusively rely on Google Cloud Storage

tormath1 opened the pull request:
In this PR, we abstract the temporary storage used by Equinix Metal tests and we bring the possibility to implement another ones (like SSH).

Related to: https://github.com/flatcar-linux/Flatcar/issues/720

## How to use

```bash
./kola run --platform=equinixmetal ... --equinixmetal-storage-url="ssh+https://my-server"
``` 

## Testing done
(Voluntary obfuscated the server name) 
```
=== RUN   bpf.execsnoop
2022-06-24T09:09:18Z platform/api/equinixmetal/storage/remote: mantle-tormath1-k124-33001fe726-02888b3dd2.ign uploaded to /home/core/content/mantle-tormath1-k124-33001fe726-02888b3dd2.ign
2022-06-24T09:09:18Z platform/api/equinixmetal: user-data available at http://my-server/mantle-tormath1-k124-33001fe726-02888b3dd2.ign
2022-06-24T09:09:18Z platform/api/equinixmetal/storage/remote: mantle-tormath1-k124-33001fe726-fa062834f0.ipxe uploaded to /home/core/content/mantle-tormath1-k124-33001fe726-fa062834f0.ipxe
2022-06-24T09:09:18Z platform/api/equinixmetal: iPXE script available at http://my-server/mantle-tormath1-k124-33001fe726-fa062834f0.ipxe
2022-06-24T09:09:18Z platform/api/equinixmetal: Recycling is not possible, creating a new instance
2022-06-24T09:09:25Z platform/api/equinixmetal: Created device: "43034ffb-29e5-4745-ba2c-c2da49877654"
2022-06-24T09:14:35Z platform/api/equinixmetal: Device active: "43034ffb-29e5-4745-ba2c-c2da49877654"
2022-06-24T09:15:05Z platform/api/equinixmetal: Finished installation of device: "43034ffb-29e5-4745-ba2c-c2da49877654"
2022-06-24T09:15:05Z platform/api/equinixmetal/storage/remote: /home/core/content/mantle-tormath1-k124-33001fe726-fa062834f0.ipxe deleted from remote storage
2022-06-24T09:15:06Z platform/api/equinixmetal/storage/remote: /home/core/content/mantle-tormath1-k124-33001fe726-02888b3dd2.ign deleted from remote storage
```

Also tested with regular GCS (default behavior)

- [x] Changelog entries added in the respective `changelog/` directory (user-facing change, bug fix, security fix, update)

jepio commented on platform/api/equinixmetal/api.go:
```diff
@@ -95,27 +96,63 @@ type Options struct {
+// created for a test.
+type Storage interface {
+	// Upload the []byte data to the implemented storage.
+	Upload(string, string, []byte) (string, string, error)
```
This could use a bit of doc about what the string parameters and returns are expected to mean.

jepio replied:
Or can one add names?

pothos commented on platform/api/equinixmetal/api.go:
```diff
@@ -158,21 +173,74 @@ func New(opts *Options) (*API, error) {
+			protocol = proto
+		}
+
+		storage = remote.NewRemote(client, url.Hostname(), opts.RemoteDocumentRoot, protocol)
```
I would expect the naming of module and constructor to be `sshStorage.NewSshStorage(`

pothos commented on platform/api/equinixmetal/api.go:
```diff
@@ -158,21 +173,74 @@ func New(opts *Options) (*API, error) {
+			},
+		}
+
+		client, err := ssh.Dial("tcp", url.Host, cfg)
```
This client establishes a TCP connection but I don't see when it gets closed. The client gets passed to the `storage` struct but it doesn't have a method to close the client. I would add one and maybe do a no-op for GCS. The only problem is when to call it, because I don't see a destructor for the API struct.

pothos replied:
`platform/machine/equinixmetal/flight.go` has a `Destroy` function which could call a destroy function for the API struct but it doesn't exist?

tormath1 commented:
@pothos thanks for your review - comments have been addressed:
- added a destructor for the API (no-op operation for the GCS implementation)
- renamed `remote` to `sshstorage`
- renamed constructors to `New` for both implementation to avoid name redundancy

pothos commented on cmd/kola/options.go:
```diff
@@ -197,8 +197,11 @@ func init() {
 	sv(&kola.EquinixMetalOptions.InstallerImageCpioURL, "equinixmetal-installer-image-cpio-url", "", "EquinixMetal installer image cpio URL, (default equinixmetal-installer-image-base-url/flatcar_production_pxe_image.cpio.gz)")
 	sv(&kola.EquinixMetalOptions.ImageURL, "equinixmetal-image-url", "", "EquinixMetal image URL (default board-dependent, e.g. \"https://alpha.release.flatcar-linux.net/amd64-usr/current/flatcar_production_packet_image.bin.bz2\")")
-	sv(&kola.EquinixMetalOptions.StorageURL, "equinixmetal-storage-url", "gs://users.developer.core-os.net/"+os.Getenv("USER")+"/mantle", "Google Storage base URL for temporary uploads")
+	sv(&kola.EquinixMetalOptions.StorageURL, "equinixmetal-storage-url", "gs://users.developer.core-os.net/"+os.Getenv("USER")+"/mantle", "Storage URL for temporary uploads (supported: gs, ssh+http, ssh+https, ssh")
```
```suggestion
	sv(&kola.EquinixMetalOptions.StorageURL, "equinixmetal-storage-url", "gs://users.developer.core-os.net/"+os.Getenv("USER")+"/mantle", "Storage URL for temporary uploads (supported: gs, ssh+http, ssh+https, or ssh which defaults to https for download)")
```
